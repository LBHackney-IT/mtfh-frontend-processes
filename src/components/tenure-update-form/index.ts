export * from "./tenure-update-form";
