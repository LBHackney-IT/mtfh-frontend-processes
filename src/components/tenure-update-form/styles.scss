.mtfh-tenure-update-form__time {
  & .lbh-select {
    margin-top: 0;
  }
}
