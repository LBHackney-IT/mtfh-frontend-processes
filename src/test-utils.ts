import {
  getAssetV1,
  getPersonV1,
  getProcessV1,
  getTenureV1,
  mockProcessV1,
  patchProcessV1,
  postProcessV1,
  server,
} from "@hackney/mtfh-test-utils";

beforeEach(() => {
  server.use(getTenureV1());
  server.use(getPersonV1());
  server.use(getAssetV1());
  server.use(postProcessV1());
  server.use(getProcessV1());
  server.use(patchProcessV1());
});

export const mockProcessSelectTenants = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "SelectTenants" },
};

export const mockProcessAutomatedChecksFailed = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "AutomatedChecksFailed" },
};

export const mockProcessAutomatedChecksPassed = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "AutomatedChecksPassed" },
};

export const mockProcessInvalidState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "InvalidState" },
};

export const mockManualChecksPassedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "ManualChecksPassed" },
};

export const mockManualChecksFailedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "ManualChecksFailed" },
};

export const mockBreachChecksFailedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "BreachChecksFailed" },
};

export const mockBreachChecksPassedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "BreachChecksPassed" },
};

export const mockDocumentsRequestedDes = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "DocumentsRequestedDes" },
};

export const mockProcessClosedState = {
  ...mockProcessV1,
  currentState: { ...mockProcessV1.currentState, state: "ProcessClosed" },
};

export const mockDocumentsRequestedDesAppointment = {
  ...mockProcessV1,
  currentState: {
    ...mockProcessV1.currentState,
    state: "DocumentsRequestedAppointment",
    processData: {
      formData: {
        appointmentDateTime: "2022-10-12T08:59:00.000Z",
      },
      documents: [],
    },
  },
  previousStates: [{ ...mockProcessV1.currentState, state: "DocumentsRequestedDes" }],
};

export const mockDocumentsRequestedAppointment = ({
  appointmentDateTime,
}: {
  appointmentDateTime: string;
}) => {
  return {
    ...mockProcessV1,
    currentState: {
      ...mockProcessV1.currentState,
      state: "DocumentsRequestedAppointment",
      processData: {
        formData: {
          appointmentDateTime,
        },
        documents: [],
      },
    },
  };
};

export const mockDocumentsAppointmentRescheduled = {
  ...mockProcessV1,
  currentState: {
    ...mockProcessV1.currentState,
    state: "DocumentsAppointmentRescheduled",
  },
};
