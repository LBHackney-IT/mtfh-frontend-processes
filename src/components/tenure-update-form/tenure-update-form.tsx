import React, { useState } from "react";

import { Form, Formik } from "formik";

import {
  TenureStartDateFormData,
  tenureStartDateIsValid,
  tenureStartDateSchema,
} from "../../schemas/tenure-start-date";
import { locale } from "../../services";
import { Trigger } from "../../services/processes/types";
import { getFormattedDateStr } from "../../utils/processUtil";

import { Process, editProcess } from "@mtfh/common/lib/api/process/v1";
import {
  Button,
  Center,
  Checkbox,
  DateField,
  FormGroup,
  Heading,
  InlineField,
  List,
  Spinner,
  Text,
} from "@mtfh/common/lib/components";
import { useErrorCodes } from "@mtfh/common/lib/hooks";

import "./styles.scss";

interface TenureUpdateFormProps {
  process: Process;
  mutate: () => void;
  setGlobalError: any;
}

export const TenureUpdateForm = ({
  process,
  mutate,
  setGlobalError,
}: TenureUpdateFormProps): JSX.Element => {
  const [checked, setChecked] = useState<boolean>(false);
  const errorMessages = useErrorCodes();

  if (!errorMessages) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }
  return (
    <>
      <Heading variant="h3">{`${locale.views.submitCase.nextSteps}:`}</Heading>
      <Text>{`${locale.views.tenureUpdateForm.checks.introduction},`}</Text>
      <List variant="bullets" style={{ marginLeft: "1em" }}>
        {locale.views.tenureUpdateForm.checks.items.map((item, index) => (
          <Text key={`tenure-update-form-check-${index}`}>{`${item}.`}</Text>
        ))}
      </List>
      <Formik<TenureStartDateFormData>
        initialValues={{ day: "", month: "", year: "" }}
        validateOnBlur={false}
        validateOnChange={false}
        validationSchema={tenureStartDateSchema(errorMessages)}
        validate={(values) => {
          return validate(errorMessages, values);
        }}
        onSubmit={async (values) => {
          const tenureStartDate = getFormattedDateStr(values);
          try {
            await editProcess({
              id: process.id,
              processTrigger: Trigger.UpdateTenure,
              processName: process?.processName,
              etag: process.etag || "",
              formData: {
                tenureStartDate,
              },
              documents: [],
            });
            mutate();
          } catch (e: any) {
            setGlobalError(e.response?.status || 500);
          }
        }}
      >
        {({ errors }) => {
          return (
            <Form
              noValidate
              id="request-tenure-update-form"
              className="request-tenure-update-form"
            >
              <DateFields
                dateLabel={locale.views.tenureUpdateForm.setNewTenureStartDate}
                errors={errors}
              />
              <InlineField name="hasNotifiedResident" type="checkbox">
                <Checkbox
                  id="condition"
                  checked={checked}
                  onClick={() => setChecked(!checked)}
                >
                  {locale.views.tenureUpdateForm.taskConformation}
                </Checkbox>
              </InlineField>
              <Button type="submit" disabled={!checked} style={{ width: 245 }}>
                {locale.confirm}
              </Button>
            </Form>
          );
        }}
      </Formik>
    </>
  );
};

export const validate = (errorMessages, values) => {
  if (values.year || values.month || values.day) {
    const result = tenureStartDateIsValid(errorMessages, values);
    if (result === true) {
      return {};
    }
    return { global: result };
  }
};

export const DateFields = ({
  dateLabel = "Date",
  errors,
}: {
  dateLabel?: string;
  errors?: any;
}): JSX.Element => {
  return (
    <FormGroup id="tenure-update-form-date-form-group" error={errors?.global}>
      <FormGroup
        id="tenure-update-form-date-field-form-group"
        style={{
          display: "flex",
          flexWrap: "wrap",
          alignItems: "flex-end",
          marginTop: 0,
        }}
      >
        <>
          <DateField
            id="tenure-update-form-date"
            className="mtfh-tenure-update-form__date"
            label={dateLabel}
            dayLabel=""
            monthLabel=""
            yearLabel=""
            dayProps={{ name: "day", placeholder: "dd" }}
            monthProps={{ name: "month", placeholder: "mm" }}
            yearProps={{ name: "year", placeholder: "yyyy" }}
            style={{ flex: "1 1 280px" }}
            required
          />
        </>
      </FormGroup>
    </FormGroup>
  );
};
