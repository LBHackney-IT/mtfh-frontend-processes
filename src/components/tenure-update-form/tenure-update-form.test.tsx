import React from "react";

import { getReferenceDataV1, render, server } from "@hackney/mtfh-test-utils";
import { screen, waitForElementToBeRemoved } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { mockDocumentsRequestedDes } from "../../test-utils";
import { TenureUpdateForm } from "./tenure-update-form";

describe("tenure-update-form-component", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  test("it renders tenure update form correctly", async () => {
    server.use(getReferenceDataV1({}, 200));
    const { container } = render(
      <TenureUpdateForm
        process={mockDocumentsRequestedDes}
        mutate={() => {}}
        setGlobalError={() => {}}
      />,
    );
    await waitForElementToBeRemoved(screen.queryAllByText(/Loading/));
    expect(container).toMatchSnapshot();
    expect(screen.getByText("Confirm")).toBeDisabled();
    await userEvent.click(screen.getByText("I confirm I have completed the tasks above"));
    expect(screen.getByText("Confirm")).toBeEnabled();
  });
});
