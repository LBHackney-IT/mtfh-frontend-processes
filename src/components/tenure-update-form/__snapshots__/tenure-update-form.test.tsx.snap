// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tenure-update-form-component it renders tenure update form correctly 1`] = `
<div>
  <h3
    class="lbh-heading-h3"
  >
    Next Steps:
  </h3>
  <p
    class="lbh-body-m"
  >
    As part of the tenure creation process,
  </p>
  <ul
    class="lbh-list lbh-list--bullet"
    style="margin-left: 1em;"
  >
    <li>
      <p
        class="lbh-body-m"
      >
        a Tenancy Termination is completed by the tenant.
      </p>
    </li>
    <li>
      <p
        class="lbh-body-m"
      >
        a new secure tenancy is signed by both joint tenants.
      </p>
    </li>
    <li>
      <p
        class="lbh-body-m"
      >
        both tenants understand and are provided with a copy "Your Tenancy Conditions".
      </p>
    </li>
    <li>
      <p
        class="lbh-body-m"
      >
        a photograph is taken of the new tenant.
      </p>
    </li>
  </ul>
  <form
    action="#"
    class="request-tenure-update-form"
    id="request-tenure-update-form"
    novalidate=""
  >
    <div
      class="govuk-form-group lbh-form-group"
      id="tenure-update-form-date-form-group"
    >
      <label
        class="govuk-label lbh-label"
        for="tenure-update-form-date-form-group-field"
      >
        
      </label>
      <div
        class="govuk-form-group lbh-form-group"
        id="tenure-update-form-date-field-form-group"
        style="display: flex; flex-wrap: wrap; align-items: flex-end; margin-top: 0px;"
      >
        <label
          class="govuk-label lbh-label"
          for="tenure-update-form-date-field-form-group-field"
        >
          
        </label>
        <fieldset
          class="govuk-form-group lbh-form-group mtfh-tenure-update-form__date"
          id="tenure-update-form-date"
          style="flex: 1 1 280px;"
        >
          <legend
            class="govuk-label lbh-label"
            for="tenure-update-form-date-field"
          >
            Set the new tenure start date
            <sup
              aria-hidden="true"
            >
              *
            </sup>
          </legend>
          <div
            class="govuk-date-input lbh-date-input"
          >
            <div
              class="govuk-date-input__item"
            >
              <label
                class="govuk-label lbh-label"
                for="tenure-update-form-date-field-day"
              />
              <input
                aria-label="Day"
                aria-valuemax="31"
                aria-valuemin="1"
                class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
                name="day"
                placeholder="dd"
                required=""
                role="spinbutton"
                value=""
              />
            </div>
            <div
              class="govuk-date-input__item"
            >
              <label
                class="govuk-label lbh-label"
                for="tenure-update-form-date-field-month"
              />
              <input
                aria-label="Month"
                aria-valuemax="12"
                aria-valuemin="1"
                class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
                name="month"
                placeholder="mm"
                required=""
                role="spinbutton"
                value=""
              />
            </div>
            <div
              class="govuk-date-input__item"
            >
              <label
                class="govuk-label lbh-label"
                for="tenure-update-form-date-field-year"
              />
              <input
                aria-label="Year"
                class="govuk-input lbh-input govuk-input govuk-date-input__input govuk-input--width-4"
                name="year"
                placeholder="yyyy"
                required=""
                role="spinbutton"
                value=""
              />
            </div>
          </div>
        </fieldset>
      </div>
    </div>
    <div
      class="govuk-checkboxes__item"
    >
      <input
        class="govuk-checkboxes__input"
        id="condition"
        name="hasNotifiedResident"
        type="checkbox"
        value=""
      />
      <label
        class="govuk-label govuk-checkboxes__label"
        for="condition"
      >
        I confirm I have completed the tasks above
      </label>
    </div>
    <button
      class="govuk-button lbh-button"
      disabled=""
      style="width: 245px;"
      type="submit"
    >
      Confirm
    </button>
  </form>
</div>
`;
